### PR TITLE
Use struct on stack instead of opaque struct in opentime [obsolete]

### LINKED ISSUE #3: Manually export individual functions
Manually export individual functions, rather than using nested export "C" blocks. The export macro will be useful later when we add visibility attributes. c.f. https://github.com/PixarAnimationStudios/OpenTimelineIO/commit/5b86f248fd6c661afbb37f400b1248139e7d4984

### PR DESCRIPTION
**Pull request by KarthikRIyer**
Closes #3

### CODE REVIEW

**Comment by KarthikRIyer** [comment]
Basic changes are done. Just need to remove pointers and `XYX_destroy()` statements from rest of the code and tests, then we can merge.

**Comment by meshula** [comment]
This closes #7

File: include/copentime/optionalOpenTime.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+};
+typedef struct OptionalTimeRange OptionalTimeRange;
+
+OTIO_API OptionalTimeRange OptionalTimeRange_create(TimeRange timeRange);
+
+OTIO_API OptionalTimeRange OptionalTimeRange_create_null();
+
+OTIO_API bool OptionalTimeRange_valid(OptionalTimeRange optionalTimeRange);
+
+OTIO_API TimeRange OptionalTimeRange_value(OptionalTimeRange optionalTimeRange);
```
Comment (meshula) [inline include/copentime/optionalOpenTime.h:39]:
```suggestion
OTIO_API TimeRange OptionalTimeRange_value(OptionalTimeRange optionalTimeRange);

```

File: include/copentime/rationalTime.h [hunk 2]
```diff
@@ -4,58 +4,84 @@
...
-    double RationalTime_to_seconds(RationalTime* self);
-    const char* RationalTime_to_timecode(
-        RationalTime*        self,
-        double               rate,
+
+struct RationalTime {
+    double value;
+    double rate;
+};
+typedef struct RationalTime RationalTime;
```
Comment (meshula) [inline include/copentime/rationalTime.h:16]: This is the pattern we've been using but I think we ought to switch to the more compact form. 
```suggestion
typedef struct {
    double value;
    double rate;
} RationalTime;
```

File: include/copentime/rationalTime.h [hunk 3]
```diff
@@ -4,58 +4,84 @@
...
+OTIO_API RationalTime RationalTime_from_time_string(
+        const char *time_string, double rate, OpenTimeErrorStatus *error_status);
+
+OTIO_API int RationalTime_to_frames(RationalTime self);
+
+OTIO_API int RationalTime_to_frames_with_rate(RationalTime self, double rate);
+
+OTIO_API double RationalTime_to_seconds(RationalTime self);
+
+OTIO_API const char *RationalTime_to_timecode(
```
Comment (meshula) [inline include/copentime/rationalTime.h:68]:
```suggestion
// the returned char* must be dealloced using free()
OTIO_API const char *RationalTime_to_timecode(
```

File: include/copentime/rationalTime.h [hunk 4]
```diff
@@ -4,58 +4,84 @@
...
-    RationalTime* RationalTime_compare(RationalTime* lhs, RationalTime* rhs);
-    bool         RationalTime_equal(RationalTime* lhs, RationalTime* rhs);
-    bool         RationalTime_not_equal(RationalTime* lhs, RationalTime* rhs);
-    void          RationalTime_destroy(RationalTime* self);
-#ifdef __cplusplus
-}
-#endif
+        OpenTimeErrorStatus *error_status);
+
+OTIO_API const char *RationalTime_to_timecode_auto(
```
Comment (meshula) [inline include/copentime/rationalTime.h:74]:
```suggestion
// The returned char* must be dealloced with free()
OTIO_API const char *RationalTime_to_timecode_auto(
```

File: include/copentime/rationalTime.h [hunk 5]
```diff
@@ -4,58 +4,84 @@
...
+        RationalTime self, OpenTimeErrorStatus *error_status);
+
+OTIO_API const char *RationalTime_to_time_string(RationalTime self);
```
Comment (meshula) [inline include/copentime/rationalTime.h:77]:
```suggestion
// The returned char* must be dealloced with free()
OTIO_API const char *RationalTime_to_time_string(RationalTime self);
```

File: include/copentime/timeRange.h [hunk 6]
```diff
@@ -1,39 +1,61 @@
...
+# define OTIO_API extern "C"
+#else
+# define OTIO_API
 #endif
+
+struct TimeRange {
+    RationalTime start_time;
+    RationalTime duration;
+};
+typedef struct TimeRange TimeRange;
```
Comment (meshula) [inline include/copentime/timeRange.h:16]:
```suggestion
typedef struct {
    RationalTime start_time;
    RationalTime duration;
} TimeRange;
```

File: include/copentime/timeTransform.h [hunk 7]
```diff
@@ -1,28 +1,41 @@
...
+# define OTIO_API
 #endif
+
+struct TimeTransform {
+    RationalTime offset;
+    double scale;
+    double rate;
+};
+typedef struct TimeTransform TimeTransform;
+
```
Comment (meshula) [inline include/copentime/timeTransform.h:19]:
```suggestion
typedef struct {
    RationalTime offset;
    double scale;
    double rate;
} TimeTransform;

```

File: include/copentime/timeTransform.h [hunk 8]
```diff
@@ -1,28 +1,41 @@
...
+OTIO_API TimeTransform
+TimeTransform_applied_to_time_transform(TimeTransform self, TimeTransform other);
+
+OTIO_API RationalTime
+TimeTransform_applied_to_rational_time(TimeTransform self, RationalTime other);
+
+OTIO_API bool TimeTransform_equal(TimeTransform lhs, TimeTransform rhs);
+
+OTIO_API bool TimeTransform_not_equal(TimeTransform lhs, TimeTransform rhs);
```
Comment (meshula) [inline include/copentime/timeTransform.h:41]:
```suggestion
OTIO_API bool TimeTransform_not_equal(TimeTransform lhs, TimeTransform rhs);

```

**Review by meshula — CHANGES_REQUESTED** [review]
- Overall, this is so much easier to read and mentally process how the API works.

- In the Swift bindings, the language specific bits are much more explicitly called out - something that does a cpp conversion or an objc conversion has those terms in the name. I think it helps overall readability and understanding of the intent.

- Would it be unreasonable to split this into two commits? One commit to introduce the structs, and another commit to change the extern C methodology? They are disjunct issues, and don't overlap much as far as I can tell. It's difficult to fully comprehend the diff because there's a mental load to determine whether a given line is merely whitespace/API, or related to the struct on stack change.

Or at least, separate all the files that don't mention time ranges but do change the ws and extern C into one big commit, and then add the rest of the files on top?

Or incrementally peel off some commits as precursors, and start rebasing this down?

The diff is so big, I hesitate to rely only on CI as a check, if it was in parts....

**Comment by KarthikRIyer** [comment]
👍 I'll make the changes.

Also I had to rename the RationalTime struct to OpenTimeRationalTime, to distinguish between it and `class RationalTime` in OTIO. That was what was giving the weird errors on Windows. Don't know why, but on making this change, it compiled.

Also, some tests are failing on Windows. Investigating it now.

**Comment by meshula** [comment]
PS, nice work!

File: include/copentimelineio/anyVector.h [hunk 9]
```diff
@@ -4,51 +4,50 @@
...
+# define OTIO_API
 #endif
+
+struct AnyVectorIterator;
+typedef struct AnyVectorIterator AnyVectorIterator;
+struct AnyVector;
+typedef struct AnyVector AnyVector;
+struct AnyVectorMutationStamp;
+typedef struct AnyVectorMutationStamp AnyVectorMutationStamp;
+OTIO_API AnyVector *AnyVector_create();
```
[the remark this answers is not in the document]
Comment (KarthikRIyer) [reply include/copentimelineio/anyVector.h:18]: CLion autoformatting did this. I'll see if I can change the settings. Or maybe we could directly add a clang-format file later?

**Comment by KarthikRIyer** [comment]
> Would it be unreasonable to split this into two commits? One commit to introduce the structs, and another commit to change the extern C methodology? They are disjunct issues, and don't overlap much as far as I can tell. It's difficult to fully comprehend the diff because there's a mental load to determine whether a given line is merely whitespace/API, or related to the struct on stack change.

> Or at least, separate all the files that don't mention time ranges but do change the ws and extern C into one big commit, and then add the rest of the files on top?

I apologise for this. I didn't pay attention to this and kept working. By the time I realised this, it was too late 😓 

I'm not sure how to split the struct part and extern C part. But separating the files mentioning opentime structs seems doable. Should I create a new PR, or force push to the same branch?

**Comment by meshula** [comment]
Could we leave this PR open for now, and create a new PR for the opentime structs? and a new PR for everything else? Then close this one, when the two new ones are known to properly encompass the work here. I would find that less confusing to review 🥰

**Comment by KarthikRIyer** [comment]
Ok, I'll do that.

Also, I think I might need some help with getting it running on Windows. It compiled after changing the struct name, but now some of the tests are failing.

**Comment by meshula** [comment]
feel free to ping me on slack with the failures you are seeing

**Comment by KarthikRIyer** [comment]
https://github.com/OpenTimelineIO/OpenTimelineIO-C-Bindings/pull/11
New PR with just the files which mention opentime structs. It still looks very big.

**Comment by meshula** [comment]
Thanks, that will help a lot! The new PR has 62 files, compared to 114 in this one.

File: include/copentime/rationalTime.h (hunk 2 above)
Comment (luke-titley) [reply include/copentime/rationalTime.h:16]: How do you forward declare RationalTime ? Might be showing my C ignorance here.

**Pull request closed without merging**